File: src/main/java/com/example/testtreemenu/service/MenuServiceImpl.java
package com.example.testtreemenu.service;

import com.example.testtreemenu.model.node.NodeMenu;
import com.example.testtreemenu.repository.MenuRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
public class MenuServiceImpl implements MenuService {
    private MenuRepository repository;

    public MenuServiceImpl(MenuRepository repository) {
        this.repository = repository;
    }

    public List<NodeMenu> getMenuTree() {
        List<NodeMenu> menuList = repository.findAll().stream().map(NodeMenu::new).toList();
        Map<Integer, NodeMenu> resultMenuMap = menuList.stream().collect(Collectors.toMap(NodeMenu::getId, Function.identity()));

        for (NodeMenu menu : menuList) {
            if (menu.getParent() != null) {
                resultMenuMap.get(menu.getParent()).addChildren(menu);
            }
        }
        for (NodeMenu menu : menuList) {
            if (menu.getParent() != null) {
                resultMenuMap.remove(menu.getId());
            }
        }

        return resultMenuMap.values().stream().toList();
    }
}

File: src/main/kotlin/com/example/treemenu/service/MenuServiceImpl.kt
package com.example.treemenu.service

import com.example.treemenu.model.node.NodeMenu
import com.example.treemenu.repository.MenuRepository
import org.springframework.stereotype.Service

@Service
class MenuServiceImpl(private val repository: MenuRepository) : MenuService {

    override val menuTree: List<NodeMenu>
        get() {
            val menuList: List<NodeMenu> = repository.findAll().map { NodeMenu(it!!) }
            val resultMenuMap: MutableMap<Int, NodeMenu> = menuList.associateBy { it.id }.toMutableMap()

            for (menu in menuList) {
                if (menu.parent != null) {
                    resultMenuMap[menu.parent]!!.addChildren(menu)
                }
            }
            for (menu in menuList) {
                if (menu.parent != null) {
                    resultMenuMap.remove(menu.id)
                }
            }
            return resultMenuMap.values.toList()
        }
}